### PR TITLE
[MODORDERS-1210] Implement API to change piece statuses in batch

### PR DESCRIPTION
**Pull request by Saba-Zedginidze-EPAM**
## Purpose
[[MODORDERS-1210] Implement API to change piece statuses in batch](https://folio-org.atlassian.net/browse/MODORDERS-1210)]

## Approach
- Minor refactoring of existing classes
- Add new endpoint
- Add status update method to PieceUpdateFlowManager
- Updated unit tests
- Added new unit tests
- Updated module descriptor

## Verification
image
image
image
image

### CODE REVIEW

File: descriptors/ModuleDescriptor-template.json [hunk 1]
```diff
@@ -627,6 +627,31 @@
             "invoice.invoice-lines.item.put"
           ]
         },
+        {
+          "methods": ["PUT"],
+          "pathPattern": "/orders/pieces-batch/status",
```
Comment (SerhiiNosko) [inline descriptors/ModuleDescriptor-template.json:632]: also please create PR in acq-models with new schemas

Comment (SerhiiNosko) [reply descriptors/ModuleDescriptor-template.json:632]: regarding enpoint name, maybe /orders/pieces-status ?

Comment (SerhiiNosko) [reply descriptors/ModuleDescriptor-template.json:632]: or idea was to support updating other fields except status in bathc, okay makes sense, so in this case lets ignore my comment

File: src/main/java/org/folio/orders/events/handlers/ReceiptStatusConsistency.java [hunk 2]
```diff
@@ -1,162 +1,85 @@
...
-            promise.fail(e);
-          }))
-      .onFailure(e -> {
-        logger.error("The error happened getting all pieces by poLine {}", poLineIdUpdate, e);
-        promise.fail(e);
-      });
 
-    // Now wait for all operations to be completed and send reply
-    completeAllFutures(futures, message);
-  }
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/orders/events/handlers/ReceiptStatusConsistency.java:109]: how simpler it become without promises, nice

File: src/main/java/org/folio/orders/events/handlers/ReceiptStatusConsistency.java [hunk 3]
```diff
@@ -1,162 +1,85 @@
...
+    completeAllFutures(List.of(future), message);
   }
 
-  private ReceiptStatus calculatePoLineReceiptStatus(PoLine poLine, List<Piece> pieces) {
-
-    if (CollectionUtils.isEmpty(pieces)) {
-      logger.info("No pieces processed - receipt status unchanged for PO Line '{}'", poLine.getId());
-      return poLine.getReceiptStatus();
+  private Future<Void> updatePoLineAndOrderStatuses(List<Piece> pieces, PoLine poLine, RequestContext requestContext) {
+    if (PoLineCommonUtil.isCancelledOrOngoingStatus(poLine)) {
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/orders/events/handlers/ReceiptStatusConsistency.java:61]: can improve by adding info log here also

File: src/main/java/org/folio/service/pieces/flows/update/PieceUpdateFlowManager.java [hunk 4]
```diff
@@ -81,78 +82,96 @@ public Future<Void> updatePiece(Piece pieceToUpdate, boolean createItem, boolean
...
-    Map<String, Piece.ReceivingStatus> map = new HashMap<>();
-    fromStorage.forEach(piece -> map.put(piece.getId(), piece.getReceivingStatus()));
-    toUpdate.forEach(piece -> map.put(piece.getId(), piece.getReceivingStatus()));
-
-    // 2. count received and expected statuses
-    long receivedQuantity = map.values().stream().filter(RECEIVED_STATUSES::contains).count();
-    long expectedQuantity = map.values().stream().filter(EXPECTED_STATUSES::contains).count();
+  private Future<Boolean> updatePiecesStatusesByPoLine(PieceBatchStatusUpdateHolder holder, RequestContext requestContext) {
+    var isAnyPiecesUpdated = holder.getPieces().stream().anyMatch(piece -> updatePieceStatus(piece, piece.getReceivingStatus(), holder.getReceivingStatus()));
+    var updates = holder.getPieces().stream().map(piece -> pieceStorageService.updatePiece(piece, requestContext)).toList();
```
[the remark this answers is not in the document]
Comment (SerhiiNosko) [reply src/main/java/org/folio/service/pieces/flows/update/PieceUpdateFlowManager.java:151]: also if isAnyPiecesUpdated = false, we can skip sending request to update pieces

[the remark this answers is not in the document]
Comment (Saba-Zedginidze-EPAM) [reply src/main/java/org/folio/service/pieces/flows/update/PieceUpdateFlowManager.java:151]: Good suggestion, fixed now

File: src/main/java/org/folio/orders/events/handlers/ReceiptStatusConsistency.java (hunk 3 above)
Comment (Saba-Zedginidze-EPAM) [reply src/main/java/org/folio/orders/events/handlers/ReceiptStatusConsistency.java:61]: Thanks, added bunch of logs

File: src/main/java/org/folio/models/pieces/PieceBatchStatusUpdateHolder.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+public class PieceBatchStatusUpdateHolder extends BasePieceFlowHolder {
+
+  private Piece.ReceivingStatus receivingStatus;
+  private List<Piece> pieces;
+  private String poLineId;
+
+  @Override
+  public String getOrderLineId() {
+    return poLineId;
+  }
```
[the remark this answers is not in the document]
Comment (Saba-Zedginidze-EPAM) [reply src/main/java/org/folio/models/pieces/PieceBatchStatusUpdateHolder.java:21]: Replaced with field @Getters

File: src/main/java/org/folio/orders/events/handlers/ReceiptStatusConsistency.java (hunk 2 above)
Comment (Saba-Zedginidze-EPAM) [reply src/main/java/org/folio/orders/events/handlers/ReceiptStatusConsistency.java:109]: Yes, I suggest we refactor older code little by little in PRs as we will have much cleaner codebase

File: src/main/java/org/folio/service/pieces/flows/update/PieceUpdateFlowManager.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (SerhiiNosko) [reply src/main/java/org/folio/service/pieces/flows/update/PieceUpdateFlowManager.java:151]: New story to introduce batch endpoint in orders-storage is created: https://folio-org.atlassian.net/browse/MODORDSTOR-435
After it completion we can use this new enpoint in this place

**Review by SerhiiNosko — APPROVED** [review]

**Review by azizbekxm — APPROVED** [review]

**Pull request merged**